File: src/main/java/com/fengsheng/handler/die_give_card_tos.java
package com.fengsheng.handler;

import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.card.Card;
import com.fengsheng.phase.AfterDieGiveCard;
import com.fengsheng.phase.WaitForDieGiveCard;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.List;

public class die_give_card_tos extends AbstractProtoHandler<Fengsheng.die_give_card_tos> {
    private static final Logger log = Logger.getLogger(die_give_card_tos.class);

    @Override
    protected void handle0(HumanPlayer r, Fengsheng.die_give_card_tos pb) {
        if (!r.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + r.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        if (!(r.getGame().getFsm() instanceof WaitForDieGiveCard fsm) || r != fsm.diedQueue.get(fsm.diedIndex)) {
            log.error("你没有死亡");
            return;
        }
        if (pb.getTargetPlayerId() == 0) {
            r.incrSeq();
            r.getGame().resolve(new AfterDieGiveCard(fsm));
        } else if (pb.getTargetPlayerId() < 0 || pb.getTargetPlayerId() >= r.getGame().getPlayers().length) {
            log.error("目标错误: " + pb.getTargetPlayerId());
            return;
        }
        if (pb.getCardIdCount() == 0) {
            log.warn("参数似乎有些不对，姑且认为不给牌吧");
            r.incrSeq();
            r.getGame().resolve(new AfterDieGiveCard(fsm));
        }
        List<Card> cards = new ArrayList<>();
        for (int cardId : pb.getCardIdList()) {
            Card card = r.findCard(cardId);
            if (card == null) {
                log.error("没有这张牌");
                return;
            }
            cards.add(card);
        }
        for (Card card : cards) r.deleteCard(card.getId());
        Player target = r.getGame().getPlayers()[r.getAbstractLocation(pb.getTargetPlayerId())];
        target.addCard(cards.toArray(new Card[0]));
        log.info(r + "给了" + target + cards);
        for (Player p : r.getGame().getPlayers()) {
            if (p instanceof HumanPlayer player) {
                var builder = Fengsheng.notify_die_give_card_toc.newBuilder();
                builder.setPlayerId(p.getAlternativeLocation(r.location()));
                builder.setTargetPlayerId(p.getAlternativeLocation(target.location()));
                if (p == r || p == target)
                    for (Card card : cards) builder.addCard(card.toPbCard());
                else
                    builder.setUnknownCardCount(cards.size());
                player.send(builder.build());
            }
        }
        r.incrSeq();
        r.getGame().resolve(new AfterDieGiveCard(fsm));
    }
}

File: src/main/kotlin/handler/die_give_card_tos.kt
package com.fengsheng.handler

import com.fengsheng.GiveCardEvent
import com.fengsheng.HumanPlayer
import com.fengsheng.card.Card
import com.fengsheng.phase.AfterDieGiveCard
import com.fengsheng.phase.WaitForDieGiveCard
import com.fengsheng.protos.Fengsheng
import com.fengsheng.protos.notifyDieGiveCardToc
import com.fengsheng.send
import org.apache.logging.log4j.kotlin.logger

class die_give_card_tos : AbstractProtoHandler<Fengsheng.die_give_card_tos>() {
    override fun handle0(r: HumanPlayer, pb: Fengsheng.die_give_card_tos) {
        if (!r.checkSeq(pb.seq)) {
            logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        val fsm = r.game!!.fsm as? WaitForDieGiveCard
        if (r !== fsm?.diedQueue?.get(fsm.diedIndex)) {
            logger.error("你没有死亡")
            r.sendErrorMessage("你没有死亡")
            return
        }
        if (pb.targetPlayerId == 0) {
            r.incrSeq()
            r.game!!.resolve(AfterDieGiveCard(fsm))
            return
        } else if (pb.targetPlayerId < 0 || pb.targetPlayerId >= r.game!!.players.size) {
            logger.error("目标错误: ${pb.targetPlayerId}")
            r.sendErrorMessage("目标错误: ${pb.targetPlayerId}")
            return
        }
        if (HashSet(pb.cardIdList).size != pb.cardIdList.size) {
            logger.error("卡牌重复${pb.cardIdList.joinToString()}")
            r.sendErrorMessage("卡牌重复${pb.cardIdList.joinToString()}")
            return
        }
        if (pb.cardIdCount == 0) {
            logger.warn("参数似乎有些不对，姑且认为不给牌吧")
            r.sendErrorMessage("参数似乎有些不对，姑且认为不给牌吧")
            r.incrSeq()
            r.game!!.resolve(AfterDieGiveCard(fsm))
            return
        }
        val cards = ArrayList<Card>()
        for (cardId in pb.cardIdList) {
            val card = r.findCard(cardId)
            if (card == null) {
                logger.error("没有这张牌")
                r.sendErrorMessage("没有这张牌")
                return
            }
            cards.add(card)
        }
        for (card in cards) r.deleteCard(card.id)
        val target = r.game!!.players[r.getAbstractLocation(pb.targetPlayerId)]!!
        if (!target.alive) {
            logger.error("目标已死亡")
            r.sendErrorMessage("目标已死亡")
            return
        }
        target.cards.addAll(cards)
        logger.info("${r}给了${target}$${cards.joinToString()}")
        r.game!!.players.send { p ->
            notifyDieGiveCardToc {
                playerId = p.getAlternativeLocation(r.location)
                targetPlayerId = p.getAlternativeLocation(target.location)
                if (p === r || p === target) cards.forEach { this.card.add(it.toPbCard()) }
                else unknownCardCount = cards.size
            }
        }
        r.incrSeq()
        r.game!!.addEvent(GiveCardEvent(fsm.whoseTurn, r, target))
        r.game!!.resolve(AfterDieGiveCard(fsm))
    }
}